### DISCUSSION TITLE
Debugging SPFx Solutions in Github Codespaces

### DISCUSSION THREAD

**Discussion by shurick81** [SharePoint/sp-dev-docs#7758] [Show and tell]
### What type of issue is this?

Question

### What SharePoint development model, framework, SDK or API is this about?

💥 SharePoint Framework

### Target SharePoint environment

SharePoint Online

### What browser(s) / client(s) have you tested

- [ ] 💥 Internet Explorer
- [ ] 💥 Microsoft Edge
- [X] 💥 Google Chrome
- [ ] 💥 FireFox
- [ ] 💥 Safari
- [ ] mobile (iOS/iPadOS)
- [ ] mobile (Android)
- [ ] not applicable
- [ ] other (enter in the "Additional environment details" area below)

### Additional environment details

- SPFx 1.13.1
- Node.js 14

### Issue description

Would it be possible to debug SPFx solutions using Github workspaces?

When I just try to run the solution as is, I get an error in the hosted workbench:

https://steria.sharepoint.com/_layouts/15/workbench.aspx?debug=true&noredir=true&debugManifestsFile=https://shurick81-alexsapozhkov-codespaces-demo-00-77p5r5x73xrpw-4321.githubpreview.dev/temp/manifests.js

![image](https://user-images.githubusercontent.com/11289124/153873860-b9013087-7540-4346-b7d4-7ab836d7ce85.png)

I try to dig deeper and checking https://shurick81-alexsapozhkov-codespaces-demo-00-77p5r5x73xrpw-4321.githubpreview.dev/temp/manifests.js in the browser, I'm getting a Bad Gateway error:

![image](https://user-images.githubusercontent.com/11289124/153873611-683e34d5-37d1-4a17-a831-2888951a3d28.png)

Then I try to fix the latter one and do the following changes (that helps for running SPFx gulp serve in containers: https://github.com/pnp/docker-spfx#known-issues):

1. Modify `node_modules\@microsoft\spfx-heft-plugins\lib\plugins\webpackConfigurationPlugin\WebpackConfigurationGenerator.js:393`

```javascript
const debugBaseUrl = `${serveConfig.https ? 'https' : 'http'}://${serveConfig.hostname || 'localhost'}:${serveConfig.port || 4321}/${distFolderName}/`;
```

to:

```javascript
const debugBaseUrl = `${serveConfig.https ? 'https' : 'http'}://localhost:${serveConfig.port || 4321}/${distFolderName}/`;
```

2. map the workbench to `0.0.0.0`, by modifying the `./config/serve.json` file in your SharePoint Framework project to:

```json
{
  "$schema": "https://developer.microsoft.com/json-schemas/core-build/serve.schema.json",
  "port": 4321,
  "hostname": "0.0.0.0",
  "https": true,
  "initialPage": "https://enter-your-SharePoint-site/_layouts/workbench.aspx"
}
```

I get the response when I request js file then:
![image](https://user-images.githubusercontent.com/11289124/153876527-99263eb9-89d9-4563-89a4-44640fb710bf.png)

However, when I try to run the hosted workbench, I get new issues:

![image](https://user-images.githubusercontent.com/11289124/153878562-dbf3e18b-5bce-466b-97e5-feb2333fdd66.png)

When I look into developer tools in the browser, I can see that the browser is getting proper response from https://shurick81-alexsapozhkov-codespaces-demo-00-77p5r5x73xrpw-4321.githubpreview.dev/temp/manifests.js but then it is trying to access resources from https://localhost:4321/temp/manifests.js and of course fails:

![image](https://user-images.githubusercontent.com/11289124/153879056-d4e6038a-ebfe-408c-99b8-0b97db70c473.png)

I might guess that the reason is that the file https://shurick81-alexsapozhkov-codespaces-demo-00-77p5r5x73xrpw-4321.githubpreview.dev/temp/manifests.js actually contains a lot of references on the localhost:

![image](https://user-images.githubusercontent.com/11289124/153879212-e81f0030-a612-44e4-80b9-5f3e06914df8.png)

**Comment by ghost** [comment]
Thank you for reporting this issue. We will be triaging your incoming issue as soon as possible.

**Comment by AJIXuMuK** [comment]
Hi @shurick81 - in 1.14 beta (that should be released soon) we introduced new configuration called `ipAddress`. This property is used explicitly wind up the server and doesn't impact the debugging urls.
Although, it was done for Docker scenarios, it might be helpful here as well. You can specify both `ipAddress` and `hostname` to achieve your needs.
Please, share the results with us, too!

**Comment by DonKirkham** [comment]
When you open an SPFx project in Codespaces, you are getting a new VM based on the default Linux container (unless you have configured .devcontainers). Regardless, because this is a new "machine," you have to generate the developer certificate, download it to your local machine and install it in the Certificate store.  If not, localhost will not deliver the manifest.js file to your browser and you get the initial errors you are seeing. (Error loading manifest.js)

**Comment by shurick81** [comment]
@DonKirkham, thanks, but as I noted above, getting manifest.js was not prevented by certificate, but only by the host name configuration. I'm not sure how it was in earlier times of Codespaces, but right now its proxy is exposing SSL with it's own certificate, so adding the certificate from container to the client device should not be necessary.

![image](https://user-images.githubusercontent.com/11289124/154426014-525fb01e-1082-4677-8f30-7d696c820c0e.png)

**Reply by DonKirkham** [reply]
I didn't realize you were attempting this from the browser. My bad. I hope you get this working as I gave up a while back. I open the Codespace in VS Code on my device and then it works just like Docker or WSL.

Hats off for attacking this. I hope we can get it solved.

**Comment by shurick81** [comment]
I have run a few tests with a new project, scaffolded with SPFx 1.14.0 that was released just today :).

I tested running in Github Codespaces a few different configurations in /config/serve.json

1. "ipAddress": "0.0.0.0"

```
{
  "$schema": "https://developer.microsoft.com/json-schemas/core-build/serve.schema.json",
  "port": 4321,
  "ipAddress": "0.0.0.0",
  "https": true,
  "initialPage": "https://steria.sharepoint.com/_layouts/15/workbench.aspx"
}
```

This worked just fine in a local container or in the Visual Studio Code Remote Container but when I try to use it with Github Codespaces, I still get this behavior: I can get /temp/manifests.js file in the browser on https://shurick81-alexsapozhkov-codespaces-demo-01-96j5q5x6h7xgq-4321.githubpreview.dev/temp/manifests.js but it is full of "https://localhost" references in the content:

![image](https://user-images.githubusercontent.com/11289124/154579459-d1dbb786-9711-48c3-8e40-ee3cfca61424.png)

And the hosted workbench throws the error, probably because of this localhost reference:

![image](https://user-images.githubusercontent.com/11289124/154579864-f5f18219-1da4-4d79-991d-31702e50219f.png)

I noticed additional issue after that. If I stop `gulp serve` and immediately run it again, I get 502 error on https://shurick81-alexsapozhkov-codespaces-demo-01-96j5q5x6h7xgq-4321.githubpreview.dev/temp/manifests.js:

![image](https://user-images.githubusercontent.com/11289124/154582547-beb870d7-5a69-464a-ac11-d73ad2ee7876.png)

This additional problem is only solved after I rebuild the whole container in Github Codespaces.

2. "hostname": "shurick81-alexsapozhkov-codespaces-demo-01-96j5q5x6h7xgq-4321.githubpreview.dev"

```
{
  "$schema": "https://developer.microsoft.com/json-schemas/core-build/serve.schema.json",
  "port": 4321,
  "hostname": "shurick81-alexsapozhkov-codespaces-demo-01-96j5q5x6h7xgq-4321.githubpreview.dev",
  "ipAddress": "0.0.0.0",
  "https": true,
  "initialPage": "https://steria.sharepoint.com/_layouts/15/workbench.aspx"
}
```

Behavior is the same as in case 1.


3. "hostname": "*"

```
{
  "$schema": "https://developer.microsoft.com/json-schemas/core-build/serve.schema.json",
  "port": 4321,
  "hostname": "*",
  "ipAddress": "0.0.0.0",
  "https": true,
  "initialPage": "https://steria.sharepoint.com/_layouts/15/workbench.aspx"
}
```

Behavior is the same as in case 1.


# Additional info

Here's the command that I used for scaffolding:

```
yo @microsoft/sharepoint --solution-name helloworld --component-type webpart --component-name hello-world-webpart --component-description "HelloWorld web part" --is-domain-isolated false --framework none --environment spo --skip-feature-deployment false
```

This is the repo:

https://github.com/alexsapozhkov/codespaces-demo-01

**Comment by shurick81** [comment]
Do I get it right that in order to make Github Codespaces web site `https://shurick81-alexsapozhkov-codespaces-demo-01-96j5q5x6h7xgq-4321.githubpreview.dev` to host assets, we should somehow make sure that on this picture, https://user-images.githubusercontent.com/11289124/154579459-d1dbb786-9711-48c3-8e40-ee3cfca61424.png, instead of `https://localhost`, the `manifests.js` file should refer to `https://shurick81-alexsapozhkov-codespaces-demo-01-96j5q5x6h7xgq-4321.githubpreview.dev`? Or is it completely wrong assumption

**Comment by anoopt** [comment]
@shurick81 - Can you please try to 
- Open the code in VS code using the **Remote explorer** extension
![image](https://user-images.githubusercontent.com/9694225/154683995-61f6aecf-0f70-4f19-9450-4d59015a35cf.png)
- In the VS code terminal run `gulp serve`
- You should see a message that port 4321 is forwarded
- In case you see `Error loading debug manifests` message - check if the manifest.js is being served from localhost first and then you can solve that error using [this post](https://www.donkirkham.com/blog/spfx-dev-cert/)
- Or you can run the below commands in VS code terminal and install the certificate on your local machine
```.sh
gulp trust-dev-cert
cp ~/.gcb-serve-data/gcb-serve.cer ./spfx-dev-cert.cer
cp ~/.gcb-serve-data/gcb-serve.cer ./spfx-dev-cert.pem
```

**Comment by shurick81** [comment]
Ok, I run "View: Show Remote Explorer"

![image](https://user-images.githubusercontent.com/11289124/154758944-ebb36a85-25de-4cad-87cf-5ce1769955bb.png)

![image](https://user-images.githubusercontent.com/11289124/154759108-c1fddfad-e38b-4591-a289-b84bc8d31bad.png)

in VS code terminal I run `gulp serve`

![image](https://user-images.githubusercontent.com/11289124/154762287-c633eb48-e008-4a39-8d76-8db2e458dd2f.png)

Where should I see that message that port 4321 is forwarded?

I can see this if I switch tab to "PORTS":

![image](https://user-images.githubusercontent.com/11289124/154762811-d4ea0e78-af62-47a4-89ab-2db0659c50f8.png)

I can see that the manifests file is served:

![image](https://user-images.githubusercontent.com/11289124/154762979-994d8fc0-89de-44e8-9faa-b613c967ba04.png)

What certificates on what local machine? On my local machine when I access manifests, they are served behind the application proxy that is using a publicly valid certificate:

![image](https://user-images.githubusercontent.com/11289124/154763135-5d9b6f82-0f28-4419-97b5-a8eb70f5cb93.png)

**Comment by anoopt** [comment]
Interesting. I think I have seen this in the past when I used the `spfx` docker image to create the codespace. Can you try without the `spfx` docker image?

- Create an empty repo in GitHub with just a README file
- Create a codespace for that repo in github.com
- Open that codespace in VS code
- In the terminal, run the command `npm install gulp-cli yo @microsoft/generator-sharepoint --global` to install the SPFx pre-reqs
- After that create a new SPFx project with `yo @microsoft/sharepoint`
- Once complete run `gulp serve`
- You should the below message in VS code at the bottom right
![image](https://user-images.githubusercontent.com/9694225/154772151-2a8a79c7-0405-45a4-b6dd-1e2831a9330c.png)
- In the **Ports** tab you should see ports `4321` and `35729` being forwarded to `127.0.0.1`
![image](https://user-images.githubusercontent.com/9694225/154772206-f4e4c575-46c1-4a9d-a05d-1cc255746797.png)

**Comment by shurick81** [comment]
The picture you are showing with ports @anoopt, it is not from Codespaces, it is from remote containers, right? In this case yes it will be 127.0.0.1 on your machine and it will be exposed via the same SSL certificate that you set up in browser. Github Codes Codespaces is different in this sense: it is exposed on the remote host. That is why it is important that manifests refer to the remote host name too. You can see in my previous comments that probably the issue is that even though all assets are hosted on the shurick81-alexsapozhkov-codespaces-demo-01-96j5q5x6h7xgq-4321.githubpreview.dev host name, manifests are referring to localhost. As a result, the browser can reach manifests file without issues but then browser executes it and trying to reach resources on localhost, which will never work for Github Codespaces.

**Comment by anoopt** [comment]
Thanks @shurick81 - the ports are from codespaces as you can see it under `Origin` column in the screenshot.

What I have noticed is when I have the `.devcontainer` folder with the files that you have in [your repo](https://github.com/alexsapozhkov/codespaces-demo-01) and create a codespace for that repo, then the codespace gets created with the ``m365pnp/spfx:1.14.0` image. With that image the ports get forwarded to the codespace host and not to 127.0.0.1.

So, if you create new repo without the `.devcontainer` folder, create a codespace for that and then do the initial SPFx setup on that codespace, open the codespace on VS code on your machine, you should see the ports forwarded correctly.

Not sure if the image `m365pnp/spfx:1.14.0` in [devcontainer.json](https://github.com/alexsapozhkov/codespaces-demo-01/blob/main/.devcontainer/devcontainer.json#L4) is not being liked by codespaces. Need to check.

**Comment by shurick81** [comment]
Thanks @anoopt, you were totally right, now it is working in codespaces in VS and everything exposed on 127.0.0.1.

I just copied the same code but removed .devcontainer directory and then created a codespace "adhoc": https://github.com/alexsapozhkov/codespaces-test-03. Here's the result:

![image](https://user-images.githubusercontent.com/11289124/155236543-73942670-33da-4d72-b55f-14ce0ee4c110.png)

![image](https://user-images.githubusercontent.com/11289124/155236580-b241191d-459e-4cd8-9fca-81eb0da129df.png)

![image](https://user-images.githubusercontent.com/11289124/155236777-829ea1c3-0b8c-4c8f-bb63-954374620132.png)


And this is considering that I don't even have Docker running on this machine, just VS and browser. The container is completely remote but somehow it is exposed on the 127.0.0.1...

How will I sleep now without understanding how this exposure is controlled...

**Comment by shurick81** [comment]
I digged a bit deeper now and the only change I needed to do in order to make the port available on the localhost is this: https://github.com/alexsapozhkov/codespaces-test-04/commit/7e637207703f7098b8bd5f12e67ec7f189742c47
The image m365pnp/spfx:1.14.0 is just fine. The only problem was the configuration of ports that dictated Codespaces to forward them to `https://*.githubpreview.dev`. See more: https://code.visualstudio.com/docs/remote/devcontainerjson-reference#_port-attributes

**Comment by shurick81** [comment]
I think we could close this issue since debugging is now working with Codespaces. However, then we need to open a new issue regarding debugging without Visual Studio installed, with just browser. When I use just browser, without Visual Studio, it is of course not possible to expose the container port on the localhost, so the port is forwarded to `https://*.githubpreview.dev`.

In such scenario, gulp serve should generate assets that refer to the remote host name in *.githubpreview.dev space. But that should be probably registered as a separate issue?

**Comment by michaelmaillot** [comment]
Hi everyone,

First of all: thanks for all the investigation, really interesting walkthrough!

I've started to have a look at running a forked repo with GitHub Codespaces to test the code and encountered some issues...

First one is when using Codespaces app (provided by VS Code but as a separate instance). I got the same forwared URL as mentioned above ("reimagined-spork-blablabla-4321.app.github.dev" here).

*devcontainer.json*
```json
{
	"name": "SPFx",
    "image": "m365pnp/spfx:1.17.3",

	// Set *default* container specific settings.json values on container create.
	"customizations": {
        "vscode": {
            "settings": {
                "terminal.integrated.shell.linux": "/bin/bash"
            },
            // Add the IDs of extensions you want installed when the container is created.
            "extensions": [
                "dbaeumer.vscode-eslint",
                "ms-vscode.vscode-typescript-tslint-plugin"
            ]
        } 
	},

	// Use 'forwardPorts' to make a list of ports inside the container available locally.
	"forwardPorts": [4321, 5432, 35729],
    "portsAttributes": {
		"4321": {
			"protocol": "https",
			"label": "Manifest",
			"onAutoForward": "silent",
			"requireLocalPort": true
		},
		"35729": {
			"protocol": "https",
			"label": "LiveReload",
			"onAutoForward": "silent",
			"requireLocalPort": true
		}
	},
	"postCreateCommand": "bash .devcontainer/spfx-startup.sh",
	"remoteUser": "node"
}
```

*serve.json* has the same config as declared [here](https://github.com/alexsapozhkov/codespaces-demo-01/blob/main/config/serve.json).

Ports are declared like this:

![image](https://github.com/SharePoint/sp-dev-docs/assets/44271458/5fe8dc63-2533-44f0-98f6-7ea44b1c91be)

I've run the `gulp trust-dev-cert` command, then run the local server and I can load the manifest file with forwarded URL in my browser. But when loading in the Workbench, I got a connection refused error, which is trying to get from "https://localhost:4321".


So I've tried directly from VS Code with which I have forwarded URL to localhost:

![image](https://github.com/SharePoint/sp-dev-docs/assets/44271458/ca1903a9-96c4-4b37-8a82-192c74aba444)

But now I got a certificate issue:

![image](https://github.com/SharePoint/sp-dev-docs/assets/44271458/207a9443-ba0f-4ed8-9113-41996940e0e7)

*devcontainer.json*

```json
{
	"name": "SPFx",
    "image": "m365pnp/spfx:1.17.3",

	// Set *default* container specific settings.json values on container create.
	"customizations": {
        "vscode": {
            "settings": {
                "terminal.integrated.shell.linux": "/bin/bash"
            },
            // Add the IDs of extensions you want installed when the container is created.
            "extensions": [
                "dbaeumer.vscode-eslint",
                "ms-vscode.vscode-typescript-tslint-plugin"
            ]
        } 
	},

	// Use 'forwardPorts' to make a list of ports inside the container available locally.
	"forwardPorts": [4321, 5432, 35729],
    "portsAttributes": {
		// "4321": {
		// 	"protocol": "https",
		// 	"label": "Manifest",
		// 	"onAutoForward": "silent",
		// 	"requireLocalPort": true
		// },
		"35729": {
			"protocol": "https",
			"label": "LiveReload",
			"onAutoForward": "silent",
			"requireLocalPort": true
		}
	},
	"postCreateCommand": "bash .devcontainer/spfx-startup.sh",
	"remoteUser": "node"
}
```

I tried suggestion above (generating the .cer file, downloading it locally and installing it on my machine) without success.

Even if what suggested @anoopt above (starting from a fresh repo then generate its Codespaces) works, my concern is regarding the use case when you want to test someone's branch / repo quickly without having to install the prerequisites. We should be able to generate a Codespace from a repo and simply launch a local browser.

Does someone figured out with a similar use case or am I doing this wrong?

**Comment by shurick81** [comment]
@michaelmaillot, are you trying to use Codespaces from Visual Studio or from a web browser? I only managed to get it working from Visual Studio

**Reply by michaelmaillot** [reply]
I'm trying to use GitHub Codespaces from VS Code.

**Comment by shurick81** [comment]
I don't know why it did not work with installing certificate in your system: the error looks exactly like the certificate is not trusted. If you want, we could try troubleshoot this together.

But in general, I think you are totally right: we need this use case just available from the SPFx tool belt, it should not require these strange troubleshooting steps when one just wants to share your commit with other people in their team.

**Reply by michaelmaillot** [reply]
Thanks for confirming that!

I would love to get some help / insights about that, maybe I'm missing a little bit of something. I'm available on Twitter, Teams,... 🙂
